### ISSUE TITLE
Add Docker support for nextflow version of the pipeline

### ISSUE THREAD

**Issue by hoelzer** [hoelzer-lab/hypro#34]
@EvaFriederike this should be straightforward now that we have a working conda-based workflow. 

## Update `container.config`

There is already a `container.config`. Here, similar to the `conda.config`, you need to add containers for the different process labels. 

## Containers

Luckily, I maintain already a large collection of Docker containers under the label `nanozoo` together w/ two colleagues. You can find the versionized and tagged containers here:

https://hub.docker.com/search?q=nanozoo&type=image

Each container encapsulates a single tool and has a specific version. So it's always clear which container was used. For example:

https://hub.docker.com/r/nanozoo/mmseqs2/tags?page=1&ordering=last_updated

where you can find

```
docker pull nanozoo/mmseqs2:11.e1a1c--55acb62
```

Just for your information, the Dockerfile (which is the recipe to build this Docker container) looksm like this:

```Dockerfile
FROM continuumio/miniconda3
ENV VERSION 11.e1a1c
ENV TOOL mmseqs2

RUN apt update && apt install -y procps wget gzip pigz && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*

RUN conda config --add channels conda-forge && \
		conda config --add channels bioconda && \
		conda config --add channels default

RUN conda install $TOOL=$VERSION && conda clean -a
```

So basically, the Docker container also just installs the tool via conda :) Which is an easy way to construct a container. The advantage: all dependencies are in the container env. 

Okay, so now please browse through the collection of containers on Dockerhub and collect those you need. If a certain version or tool is not available, please let me know via this issue and I can easily construct the image.

## Finalize

We already have a docker profile in the `nextflow.config` so this should be fine already. Once the `container.config` is set rule to run the whole workflow w/ `-profile local,docker` instead of conda.

**Comment by EvaFriederike** [comment]
@hoelzer It's coming along pretty well. 
1)For the `restore` process after the prokka annotation I had issues with file access rights in the docker profile: 

- In this  process, the prokka output is unpacked,  the original contig ids are restored and the files are packed again using the same filename as for the input. This worked fine when running hypro in the conda profile. 
- When running in docker profile, the prokka output could only be accessed by root which caused a "permission denied" error in the workflow.

=> As a workaround I gave the restored prokka output a different name. This change is not very drastic but I actually just wanted to make sure if you know of this kind of issue or if maybe this came up as a result of the docker user/access/mount settings on my linux machine.

2) The docker profile is almost done. There is just one container missing for the mygene package used for updating prokka (mapping scientific names and symbols between the mmseqs2 output and input database). There is no mygene container supported by nanozoo but by four other docker repositories:
Two of them don't provide any readme or comment on their containers and the other two appear to have created modified versions of mygene. What is your suggestion how to proceed here?

**Comment by hoelzer** [comment]
@EvaFriederike 

> => As a workaround I gave the restored prokka output a different name. This change is not very drastic but I actually just wanted to make sure if you know of this kind of issue or if maybe this came up as a result of the docker user/access/mount settings on my linux machine.

Yeah it hink that's a valid workaround. And this is most likely an issue bc/ of how Docker is configured on your system. You can configure Docker to run with your $USER as the default and then this should not happen. I imagine all files e.g. in the work dir written by Docker-processes are also `root` and thus you need to remove them via `sudo`, right? 

Anyway, the workaround should be fine and also precents issues w/ other users that only configured Docker to run as root. 

>  There is just one container missing for the mygene package used for updating prokka 

Easy, I build one ;) 

nanozoo/mygene:3.2.2--6b09e53

Let me know if that one works.

**Comment by EvaFriederike** [comment]
@hoelzer the mygene container works perfectly thanks!
Now, since the process _update\_prokka_ runs as a docker process it additionally requires a container for pandas ...might as well have foreseen that ;)

**Comment by hoelzer** [comment]
@EvaFriederike :) so you need another container with which dependencies? Only

* python3
* pandas

? 

Can you try nanozoo/president:0.6.0--54d5a21 ? That one should have py3 and pandas (and some other stuff)

**Comment by hoelzer** [comment]
docker pull nanozoo/python_rnaseq:3.8--7a7808c

should also work, also has pandas installed afaik

**Comment by EvaFriederike** [comment]
@hoelzer Exactly, I just needed python3 and pandas. Both containers you suggested work.
However, I'm having problems with this process using two different containers...is this even possible or can I only use one container per process?

**Comment by hoelzer** [comment]
you need one container for one process.

just decide for one of the two :) 

or what else is needed in that process? You either need all dependencies in one container (which I can do if you let me know what's needed in the container for this process) or you need to separate the process into two

**Comment by EvaFriederike** [comment]
Ah okay then I was trying the impossible previoulsy^^
For the ``update_prokka`` process I need a container that provides with python3: mygene and pandas.

**Comment by hoelzer** [comment]
ahh k so I just add pandas to the mygene box

**Comment by hoelzer** [comment]
3.2.2--f1d3f12 should have py3.8 and pandas now (and mygene)

**Comment by EvaFriederike** [comment]
I tried 3.2.2--f1d3f12 and though it does contain python 3.8 and the mygene package, I still get a "Module Not Found Error" for pandas when calling the ``update_prokka.py`` script...:-/

**Comment by hoelzer** [comment]
@EvaFriederike hm you're right. Just tested it via 

```bash
docker run --rm -it nanozoo/mygene:3.2.2--f1d3f12 /bin/bash
(base) root@2ef51918055b:/# python
Python 3.8.3 (default, May 19 2020, 18:47:26)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import mygene
>>> import pandas as pd
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'pandas'
>>>
```

I build a new Docker where I installed `pandas` now via `pip`: `nanozoo/mygene:3.2.2--42af2bf`

should work now.

**Comment by EvaFriederike** [comment]
@hoelzer the docker version runs smoothly now. I tested both docker and conda with a single fasta and a csv of fasta file paths as input. 
For the latter one I selected an arbitrary plasmid fasta from some labs github repository (sadly I did not write down which) that I put into the test/data/ directory...maybe you have a second test file that makes more sense. For reproducing this, one just has to adjust the absolute file paths in test/input.csv.

**Comment by hoelzer** [comment]
@EvaFriederike cool. A meaningful test file would be just any bacterial genome FASTA. Maybe the famous _E. coli_ will do it: 

https://www.ncbi.nlm.nih.gov/genome/?term=E+coli

https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz

Maybe we just add/test this as well and then it's done?

**Comment by hoelzer** [comment]
Could be that we dont get so many hypothetical annotations for _E. coli_ bc/ databases are well supported. But anyway, it's more for testing and to show that the workflow runs.
